### PR TITLE
[MRG] HW4

### PR DESCRIPTION
**Pull request by kjacks21**
# Assignment
**Goal**: Rep-Invariants, contracts, tests.
Revisit the GrowList class from assignment 3. You should consider the **mutable** version. The abstraction function is already implemented, but the rep-invariant is not.

There are two parts to the assignment.

1. Implement the rep-invariant (as repOk()).
2. Consider the statement deletion mutation operator (SDL) , which deletes a statement from a program. A mutant produced by this operator is detected if there is a test that has different outputs on the original program and the mutated program. Find two instances of this operator:
    - An SDL mutant that causes the rep-invariant to fail. Demonstrate this with a call to repOk() from a (failing) JUnit test.
    - An SDL mutant that causes a method contract to be violated. Demonstrate this with a call to the relevant method from a (failing) JUnit test.

To help out the GTA, document your SDL faults with easily identifiable comments in your code. Make the two faults/tests independent, which means that you find a test that violate the rep-invariant without violating a contract and vice versa. If that is not possible, explain why not.

# Tasks
- [x]  Implement the rep-invariant (as repOk()).
- [x] An SDL mutant that causes the rep-invariant to fail. Demonstrate this with a call to repOk() from a (failing) JUnit test.
- [x] An SDL mutant that causes a method contract to be violated. Demonstrate this with a call to the relevant method from a (failing) JUnit test.
- [x] To help out the GTA, document your SDL faults with easily identifiable comments in your code. Make the two faults/tests independent, which means that you find a test that violate the rep-invariant without violating a contract and vice versa. If that is not possible, explain why not.

### CODE REVIEW

**Comment by kjacks21** [comment]
My proposed rep invariants are as follows (I can also implement the repOK):

```
// The rep invariant is
//     GrowList.values != null &&
//     all elements of GrowList.values are Strings &&
//     there are no duplicate elements in GrowList.values
```
Thoughts?

**Comment by 1000DS000000** [comment]
I'm not sure about the check for strings, since the class uses values of E. I was thinking we could check that the keys has to be Integer.

**Comment by kjacks21** [comment]
Like below?

// The rep invariant is
//     GrowList.values != null &&
//     all keys of GrowList.values are Integers

**Comment by 1000DS000000** [comment]
Yeah.

**Comment by 1000DS000000** [comment]
I'm trying to think if there are any other rep invariants, but currently i cant think of any.

**Comment by kjacks21** [comment]
Sounds good. Changes pushed 57d0d28

**Comment by 1000DS000000** [comment]
I'm also not sure why that happens. 
I'm still trying to understand the SDL operation, but am currently working on the one about the method contract.

**Comment by ngreenwald89** [comment]
I don't see how to do either of the things in the rep-invariant without causing an error--I don't think we can get the GrowList to have keys not of type integer, or how to set values to null. So we won't be able to get the repOk() function to return false. (Right now I can't test this because I can't get the code to run; when running via Intellij I get:
`Error: Could not find or load main class GrowList
Caused by: java.lang.ClassNotFoundException: GrowList`

**Comment by ngreenwald89** [comment]
@1000DS000000 both @kjacks21 and I could not find SDL mutants that would simultaneously allow the rep-invariant and method contract tests to run. To make the rep-invariant fail and cause repOk() to return false, we commented out the GrowList constructor (see latest commit). However, this screws up the GrowList methods, and thus the method contract SDL mutant throws an NPE instead of IndexOutOfBoundsException. We think the HW instructions allow for this outcome, if we clearly explain the two SDL's, their outcomes, and their independence. What do you think?

**Comment by 1000DS000000** [comment]
I think so, since they are to be independent from each other.

**Comment by ngreenwald89** [comment]
I added some small comments to the SDL. Can you guys look it over, and if you think it's ready, submit? Thanks.

File: hw4/src/GrowListTest.java [hunk 1]
```diff
@@ -0,0 +1,21 @@
+import java.awt.List;
+import static org.junit.Assert.*;
+
+import org.junit.Test;
+
+public class GrowListTest {
+    GrowList<String> g = new GrowList<String>();
+
+    // Testing the rep-invariant; correct SDL needs to be in place.
```
Comment (kjacks21) [inline hw4/src/GrowListTest.java]: Should we say here that it is not possible to break the rep invariant and not break the contract?

Comment (1000DS000000) [reply hw4/src/GrowListTest.java]: yeah, I think there is a good spot to put that comment. Other than that I think its ready to submit

**Comment by kjacks21** [comment]
Should we submit the HW with the code commented out (SDL mutants)?

**Comment by ngreenwald89** [comment]
Yes, imo.

**Comment by kjacks21** [comment]
Right then, HW4 has been submitted on BB. Thanks all

**Comment by ngreenwald89** [comment]
Thanks Kyle.

**Comment by kjacks21** [comment]
FYI feedback from grader for a total of 10.25/10:

> Removing "values = new HashMap<Integer,E>();" in the constructor does not break the contract, because there is no contract written for the constructor. (-0.25 pts)
Collaboration bonus: +0.5 pts

**Pull request merged**